### DISCUSSION TITLE
サブクエリ実装方法について

### DISCUSSION THREAD

**Discussion by CC-Tani** [domaframework/doma#960] [Q&A] [answered]
table1とteble2がある前提で、teble1に紐づいているtable2の件数、teble1に紐づいているtable2のうち特定の条件に該当する件数を取得したいと考えています。（件数＋table1データの取得）下記の2パターンで取得可能であると考えてますが、実装方法が分からない状況です。知見がございましたらご共有よろしくお願いいたします。

①外部結合で特定の件数を取得（サブクエリ）して、その情報をselect句で他データと共に取得
②select句でサブクエリを使用して、件数を取得＋別名を定義

**Comment by nakamura-to** [comment]
Criteria APIで実現したいということでしょうか？
どんなSQLを発行したいのかを示していただければCriteria APIで実現可能かどうかお答えできると思います。

**Reply by CC-Tani** [reply]
記載が漏れており申し訳ございません。おっしゃる通りCriteria APIで実現したいと考えてます。

下記のような挙動をするSQLを発行したいです。
`select
    table1.*
    , sub1.count1
    , sub2.count2 
from
    table1 
    left join ( 
        select
            count(culmun2) as count1
            , culumn1 
        from
            table2 
        group by
            table2.culumn1
    ) sub1 
        on sub1.culumn1 = table1.culumn1 
    left join ( 
        select
            count(culmun2) as count2
            , culumn1 
        from
            table2 
        where
            culumn3 = "A" 
        group by
            table2.culumn1
    ) sub2 
        on sub2.culumn1 = table1.culumn1
`

**Reply by nakamura-to** [reply, marked as answer]
SQLの共有ありがとうございます。
Criteria APIではJOIN対象にサブクエリを指定できません。
代替としてSQLファイルやSqlアノテーションを使う方法があります。

**Reply by t-eimizu** [reply]
このQ&Aに関連して、最近実装された AggregateStrategy についての質問です。
上のようなSELECT文の sub1, sub2 という名前がついたサブクエリについて、エンティティクラスを定義して、
AggregateStrategyによるマッピングは可能でしょうか？

上のSQLにはありませんが、 table1 に table4, table5 などを join して AggregateStrategy を使って取得することはできています。
（ left join table4 t4 on t4.id = t1.some_id のような）

**Reply by nakamura-to** [reply]
> このQ&Aに関連して、最近実装された AggregateStrategy についての質問です。
上のようなSELECT文の sub1, sub2 という名前がついたサブクエリについて、エンティティクラスを定義して、
AggregateStrategyによるマッピングは可能でしょうか？

できますよ。
こんな感じのコードが必要になると思います。

```java
@Entity
record Sub1(@Id Integer column1, Integer count1)

@Entity
record Sub2(@Id Integer column1, Integer count2)

@Entity
class Table1 {
  @Id Integer id;
  @Association Sub1 sub1;
  @Association Sub2 sub2;
}

@AggregateStrategy(root = Table1.class, tableAlias = "table1")
interface MyAggregateStrategy {
  @AssociationLinker(propertyPath = "sub1", tableAlias = "sub1")
  BiConsumer<Table1, Sub1> sub1 =
      (t, s) -> {
        t.sub1 = s;
      };

  @AssociationLinker(propertyPath = "sub2", tableAlias = "sub2")
  BiConsumer<Table1, Sub2> sub2 =
      (t, s) -> {
        t.sub2 = s;
      };
} 
```

注意点ですが、Sub1とSub2のエンティティクラスの実質的にユニークなプロパティには `@Id` が必要です。

**Reply by t-eimizu** [reply]
回答いただきありがとうございます。
Table1側のSub1 に対して lombok の@Getter, @Setter をつけていたのが良くなかったみたいで、
propertyPathがうまく見つけられなかったようです。
public フィールドに変更したところすんなり動きました。

**Comment by nakamura-to** [comment]
> Criteria APIではJOIN対象にサブクエリを指定できません。

以前のバージョンではできなかったのですが、最近のバージョン（2.55.0以上）ではサポートしています。

https://docs.domaframework.org/ja/latest/query-dsl/#derived-table-expression
